docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,673 @@
+
+```ts
+interface HotContext {
+  file: string
+  timestamp: number
+  environment: string
+  modules: Array<ModuleNode>
+  read: () => string | Promise<string>
+  server: ViteDevServer
+}
+```
+
+- `environment` is the module execution environment where a file update is currently being processed.
+
+- `modules` is an array of modules in this environment that are affected by the changed file. It's an array because a single file may map to multiple served modules (e.g. Vue SFCs).
+
+- `read` is an async read function that returns the content of the file. This is provided because, on some systems, the file change callback may fire too fast before the editor finishes updating the file, and direct `fs.readFile` will return empty content. The read function passed in normalizes this behavior.
+
+The hook can choose to:
+
+- Filter and narrow down the affected module list so that the HMR is more accurate.
+
+- Return an empty array and perform a full reload:
+
+  ```js
+  hotUpdate({ environment, modules, timestamp }) {
+    if (environment !== 'browser')
+      return
+
+    const serverEnvironment = server.environment(environment)
+
+    serverEnvironment.hot.send({ type: 'full-reload' })
+    // Invalidate modules manually
+    const invalidatedModules = new Set()
+    for (const mod of modules) {
+      serverEnvironment.moduleGraph.invalidateModule(
+        mod,
+        invalidatedModules,
+        timestamp,
+        true
+      )
+    }
+    return []
+  }
+  ```
+
+- Return an empty array and perform complete custom HMR handling by sending custom events to the client:
+
+  ```js
+  hotUpdate({ environment }) {
+    if (environment !== 'browser')
+      return
+
+    server.environment(environment).hot.send({
+      type: 'custom',
+      event: 'special-update',
+      data: {}
+    })
+    return []
+  }
+  ```
+
+  Client code should register the corresponding handler using the [HMR API](./api-hmr) (this could be injected by the same plugin's `transform` hook):
+
+  ```js
+  if (import.meta.hot) {
+    import.meta.hot.on('special-update', (data) => {
+      // perform custom update
+    })
+  }
+  ```
+
+### Using `apply`
+
+In the examples of the previous section, we used a guard in the `hotUpdate` hook to only process updates from the browser environment. If a plugin is specific to only some of the available environments, `apply` can be used to avoid guarding each hook and improve performance.
+
+```js
+function ssrOnlyPlugin() {
+  return {
+    name: 'ssr-only-plugin',
+    apply({ environment }) => environment === 'ssr',
+    // unguarded hooks...
+  }
+}
+```
+
+### Accessing the current environment in hooks
+
+The `environment` name is passed as a parameter to the `options` of `resolveId`, `load`, and `transform`.
```

Comment:
If we want to conditionally return a virtual file for worker-like runtimes such as  [Cloudflare workerd](https://github.com/cloudflare/workerd) and [Fastly Compute](https://www.fastly.com/documentation/guides/compute/javascript/), is there a way to do it at once?

I guess it is required to write it like:
```js
const p = {
  name: 'virtual',
  load(id, opts) {
    // it requires to write all the environment names
    if (opts.environment === 'workerd' || opts.environment === 'fastly-compute') {
      return 'a code for worker-like runtimes'
    }
    return 'a code'
  }
}
```
But to migrate from `config.ssr.target === 'webworker'`, I think it'd nice to write it something like:
```js
const p = {
  name: 'virtual',
  load(id, opts) {
    const e = server.environment(environment)
    if (e.type === 'worker') {
      return 'a code for worker-like runtimes'
    }
    return 'a code'
  }
}
```

Reply:
This is a great point. We discussed two weeks ago a bit about this with @sheremet-va. In Vite core plugins, we have several conditions that depend on the `ssr` flag and other that depend on `ssr.target: node | webworker` as you explain (other `ssr` config options are already accounted for in the current state of the proposal).

If we want to remove the `ssr` flag, we need something like you propose here. I think we should move away from conditionals based on the environment name to conditionals based on what the environment supports. So, totally agree with you. I don't think it should be called `type` as it is too generic, but we need to identify what are the conditions not covered by `resolve.conditions` et al and add them as part of the environment config.

Maybe something like:
```ts
type runtimeType = 'browser' | 'node' | 'worker' | 'edge' | ?
```

The problem with this is that there will be differences between edge environments and node environments. So it may be worth checking all conditionals in core and other projects and see if we can be even more fine-grained. Maybe it is a more generic option we should add to resolve for example.

If we end up using only fine-grained options in core, I still think that having a `runtimeType` is a good idea, even if it is only to help with debugging.

Reply:
Can you show an example when this is useful? I think having fine-grained config should be enough.

Reply:
I was thinking on the `runtimeType` also as a way for community plugins to be able to ship features before we add the proper options in Vite core. If we are missing a fine-grained config, it takes time for it to end up in a release. So it may be useful to know the runtime type to do some conditionals until they can switch to fine-grained options. I'm good avoiding this until we have a good use case to justify it.

Reply:
I assumed that we have many `ssr.target` conditions in our codebase, but it turns out those can be rewrote with a fine-grained option. Yeah, it seems we don't need something like this.
I was thinking of returning a virtual module conditionally, but I noticed that it can be achieved by using an `exports` field with `wintercg` condition.